flake.nix

```diff
@@ -0,0 +1,138 @@
+{
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
+    flake-utils.url = "github:numtide/flake-utils";
+
+    rust-overlay.url = "github:oxalica/rust-overlay";
+    rust-overlay.inputs.nixpkgs.follows = "nixpkgs";
+
+    pyproject-nix.url = "github:pyproject-nix/pyproject.nix";
+    pyproject-nix.inputs.nixpkgs.follows = "nixpkgs";
+  };
+
+  outputs = {
+    self,
+    nixpkgs,
+    flake-utils,
+    rust-overlay,
+    pyproject-nix,
+  } @ inputs:
+    flake-utils.lib.eachDefaultSystem
+    (
+      system: let
+        overlays = [(import rust-overlay)];
+        pkgs = import nixpkgs {inherit system overlays;};
+        lib = pkgs.lib;
+
+        # If we don't do this, the renderers and validators will use the old project attrset
+        updateProjectMethods = let
+          inherit (pyproject-nix.lib) renderers validators;
+          curryProject = attrs: project:
+            lib.mapAttrs (
+              _: func: args:
+                func (args // {inherit project;})
+            )
+            attrs;
+        in
+          project:
+            (lib.makeExtensible (final: project)).extend (
+              final: prev: {
+                # Set renderers/validators to use new project deps
+                renderers = curryProject renderers final;
+                validators = curryProject validators final;
+              }
+            );
+
+        # We need to remove the "maturin" dependency because it is installed via nix, and not in the python packages
+        removeMaturinDeps = project: let
+          filteredDeps = lib.mapAttrs (name: value:
+            if lib.elem name ["dependencies" "build-systems"]
+            then lib.filter (dep: dep.name != "maturin") value
+            else value)
+          project.dependencies;
+
+          project' = lib.setAttr project "dependencies" filteredDeps;
+        in
+          updateProjectMethods project';
+
+        pyproject = removeMaturinDeps (pyproject-nix.lib.project.loadPyproject {projectRoot = ./.;});
+
+        rust = pkgs.rust-bin.stable.latest.default;
+        rustPlatform = pkgs.makeRustPlatform {
+          rustc = rust;
+          cargo = rust;
+        };
+
+        python = pkgs.python3.override {
+          self = python;
+          packageOverrides = pyfinal: pyprev: {
+            pydantic-settings = pyprev.pydantic-settings.overridePythonAttrs (old: rec {
+              version = "2.6.1";
```

Comment:
Why are versions being overriden?

Reply:
The default versions are too old, this just sets then to whats in the pyproject.toml